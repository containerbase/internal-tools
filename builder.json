{
  "datasource": "npm",
  "startVersion": "1.12.0",
  "versions": ["1.9.0", "1.12.2", "2.0.0-rc.24", "2.0.0-rc.27"]
}
