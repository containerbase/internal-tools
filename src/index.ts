// Node v16 is missing structuredClone, so we polyfill it here.
import 'core-js/actual/structured-clone';
import run from './runner';

run().catch(console.error);
