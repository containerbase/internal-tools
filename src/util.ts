import { exec as _exec } from '@actions/exec';
import { ExecOptions as _ExecOptions } from '@actions/exec/lib/interfaces';
import log from './utils/logger';
import { getInput, startGroup, endGroup } from '@actions/core';
import { join } from 'path';
import fs from 'fs';
import { promisify } from 'util';
import findUp = require('find-up');

const readFileAsync = promisify(fs.readFile);

export type ExecOptions = _ExecOptions;

export type ExecResult = {
  readonly code: number;
  readonly stdout: string;
  readonly stderr: string;
};

export async function exec(
  cmd: string,
  args: string[],
  options?: ExecOptions
): Promise<ExecResult> {
  let stdout = '';
  let stderr = '';
  let code: number;

  try {
    startGroup(`${cmd} ${args.join(' ')}`);
    code = await _exec(cmd, args, {
      ...options,
      listeners: {
        stdout: (data: Buffer) => {
          stdout += data.toString();
        },
        stderr: (data: Buffer) => {
          stderr += data.toString();
        },
      },
    });
  } finally {
    endGroup();
  }
  if (code) {
    log.error({ code });
    throw new Error('Command failed');
  }

  return { code, stdout, stderr };
}

export function isDryRun(): boolean {
  const val = getInput('dry-run');
  return !!val && val === 'true';
}

/**
 * Get environment variable or empty string.
 * Used for easy mocking.
 * @param key variable name
 */
export function getEnv(key: string): string {
  return process.env[key] ?? '';
}

export async function readJson<T = unknown>(file: string): Promise<T> {
  const path = join(getEnv('GITHUB_WORKSPACE'), file);
  const res = await import(path);
  // istanbul ignore next
  return res?.default ?? res;
}

export async function readFile(file: string): Promise<string> {
  const path = join(getEnv('GITHUB_WORKSPACE'), file);
  return await readFileAsync(path, 'utf8');
}

export const MultiArgsSplitRe = /\s*(?:[;,]|$)\s*/;

export function getArg(name: string, opts?: { required?: boolean }): string;
export function getArg(
  name: string,
  opts?: { required?: boolean; multi: true }
): string[];
export function getArg(
  name: string,
  opts?: { required?: boolean; multi?: boolean }
): string | string[];
export function getArg(
  name: string,
  opts?: { required?: boolean; multi?: boolean }
): string | string[] {
  const val = getInput(name, opts);
  return opts?.multi ? val.split(MultiArgsSplitRe).filter(Boolean) : val;
}

let _pkg: Promise<string | undefined>;

/**
 * Resolve path for a file relative to renovate root directory (our package.json)
 * @param file a file to resolve
 */
export async function resolveFile(file: string): Promise<string> {
  if (!_pkg) {
    _pkg = findUp('package.json', { cwd: __dirname, type: 'file' });
  }
  const pkg = await _pkg;
  // istanbul ignore if
  if (!pkg) {
    throw new Error('Missing package.json');
  }
  return join(pkg, '../', file);
}
