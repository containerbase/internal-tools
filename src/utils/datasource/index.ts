import { getDatasources } from 'renovate/dist/modules/datasource';
import log from '../logger';

export function register(): void {
  log('register datasources');
  getDatasources();
}
