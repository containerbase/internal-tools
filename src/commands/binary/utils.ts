import { getInput } from '@actions/core';
import is from '@sindresorhus/is';
import { getDefaultVersioning } from 'renovate/dist/modules/datasource';
import { getArch, getArg, getDistro, isDryRun, readJson } from '../../util';
import { readDockerConfig } from '../../utils/config';
import {
  DockerPlatform,
  dockerBuildx,
  dockerRun,
} from '../../utils/docker/common';
import log from '../../utils/logger';
import type { BinaryBuilderConfig, ConfigFile } from '../../utils/types';

export async function getConfig(): Promise<BinaryBuilderConfig> {
  const configFile = getInput('config') || 'builder.json';

  const cfg = await readJson<ConfigFile>(configFile);

  if (!is.object(cfg)) {
    throw new Error('missing-config');
  }

  if (!is.string(cfg.image)) {
    cfg.image = getInput('image', { required: true });
  }

  if (!is.string(cfg.buildArg)) {
    cfg.buildArg = cfg.image.toUpperCase() + '_VERSION';
  }

  await readDockerConfig(cfg);

  return {
    ...cfg,
    ignoredVersions: cfg.ignoredVersions ?? [],
    dryRun: isDryRun(),
    lastOnly: getArg('last-only') == 'true',
    buildArgs: getArg('build-args', { multi: true }),
    versioning: cfg.versioning ?? getDefaultVersioning(cfg.datasource),
  } as BinaryBuilderConfig;
}

export async function createBuilderImage(
  ws: string,
  { buildArgs }: BinaryBuilderConfig
): Promise<void> {
  log('Creating builder image');
  const args = [
    'build',
    '--load',
    '-t',
    'builder',
    '--build-arg',
    `DISTRO=${getDistro()}`,
  ];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  if (is.nonEmptyArray(buildArgs)) {
    args.push(...buildArgs.map((b) => `--build-arg=${b}`));
  }
  await dockerBuildx(...args, ws);
}

export async function runBuilder(ws: string, version: string): Promise<void> {
  const args = ['--name', 'builder', '--volume', `${ws}/.cache:/cache`];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  await dockerRun(...args, 'builder', version);
}
