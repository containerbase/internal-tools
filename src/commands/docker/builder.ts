import { getInput, setFailed } from '@actions/core';
import is from '@sindresorhus/is';
import * as chalk from 'chalk';
import { getDefaultVersioning } from 'renovate/dist/modules/datasource';
import { get as getVersioning } from 'renovate/dist/modules/versioning';
import { exec, getArg, isDryRun, readJson } from '../../util';
import { BuildsResult, addHostRule, getBuildList } from '../../utils/builds';
import { readDockerConfig } from '../../utils/config';
import { build, publish } from '../../utils/docker';
import { init } from '../../utils/docker/buildx';
import { dockerDf, dockerPrune, dockerTag } from '../../utils/docker/common';
import log from '../../utils/logger';
import type { Config, ConfigFile } from '../../utils/types';

export const MultiArgsSplitRe = /\s*(?:[;,]|$)\s*/;

function createTag(tagSuffix: string | undefined, version: string): string {
  return is.nonEmptyString(tagSuffix) && tagSuffix !== 'latest'
    ? `${version}-${tagSuffix}`
    : version;
}

async function buildAndPush(
  {
    imagePrefix,
    image,
    buildArg,
    buildArgs,
    buildOnly,
    cache,
    dryRun,
    tagSuffix,
    versioning,
    majorMinor,
    prune,
  }: Config,
  tobuild: BuildsResult
): Promise<void> {
  const builds: string[] = [];
  const failed: string[] = [];
  const ver = getVersioning(versioning);
  const versionsMap = new Map<string, string>();
  if (majorMinor) {
    for (const version of tobuild.versions) {
      const minor = ver.getMinor(version);
      const major = ver.getMajor(version);
      const isStable = ver.isStable(version);

      if (isStable && is.number(major) && `${major}` !== version) {
        versionsMap.set(`${major}`, version);
      }

      if (
        isStable &&
        is.number(major) &&
        is.number(minor) &&
        `${major}.${minor}` !== version
      ) {
        versionsMap.set(`${major}.${minor}`, version);
      }
    }
  }

  await exec('df', ['-h']);

  for (const version of tobuild.versions) {
    const tag = createTag(tagSuffix, version.replace(/\+.+/, ''));
    const imageVersion = `${imagePrefix}/${image}:${tag}`;
    log(`Building ${imageVersion}`);
    try {
      const minor = ver.getMinor(version);
      const major = ver.getMajor(version);
      const cacheTags: string[] = [tagSuffix ?? 'latest'];
      const tags: string[] = [];

      if (
        is.number(major) &&
        majorMinor &&
        versionsMap.get(`${major}`) === version
      ) {
        const nTag = createTag(tagSuffix, `${major}`);
        cacheTags.push(nTag);
        tags.push(nTag);
      }

      if (
        is.number(major) &&
        is.number(minor) &&
        majorMinor &&
        versionsMap.get(`${major}.${minor}`) === version
      ) {
        const nTag = createTag(tagSuffix, `${major}.${minor}`);
        cacheTags.push(nTag);
        tags.push(nTag);
      }

      if (version === tobuild.latestStable) {
        tags.push(tagSuffix ?? 'latest');
      }

      await build({
        image,
        imagePrefix,
        tag,
        cache,
        cacheTags,
        buildArgs: [...(buildArgs ?? []), `${buildArg}=${version}`],
        dryRun,
      });

      if (!buildOnly) {
        await publish({ image, imagePrefix, tag, dryRun });
        const source = tag;

        for (const tag of tags) {
          log(`Publish ${source} as ${tag}`);
          await dockerTag({ image, imagePrefix, src: source, tgt: tag });
          await publish({ image, imagePrefix, tag, dryRun });
        }
      }

      log(`Build ${imageVersion}`);
      builds.push(version);
    } catch (err) {
      log.error(err);
      failed.push(version);
    }

    await dockerDf();
    await exec('df', ['-h']);

    if (prune) {
      await dockerPrune();
      await exec('df', ['-h']);
    }
  }

  if (builds.length) {
    log('Build list:' + builds.join(' '));
  }

  if (failed.length) {
    log.warn('Failed list:' + failed.join(' '));
    throw new Error('failed');
  }
}

async function generateImages(config: Config): Promise<void> {
  const buildList = await getBuildList(config);

  if (!buildList?.versions.length) {
    setFailed(`No versions found.`);
    return;
  }

  await buildAndPush(config, buildList);
}

export async function run(): Promise<void> {
  const dryRun = isDryRun();
  const configFile = getInput('config') || 'builder.json';

  const cfg = await readJson<ConfigFile>(configFile);

  if (!is.object(cfg)) {
    throw new Error('missing-config');
  }

  // TODO: validation
  if (!is.string(cfg.image)) {
    cfg.image = getInput('image', { required: true });
  }

  if (!is.string(cfg.buildArg)) {
    cfg.buildArg = cfg.image.toUpperCase() + '_VERSION';
  }

  await readDockerConfig(cfg);

  const config: Config = {
    ...cfg,
    imagePrefix: getArg('image-prefix')?.replace(/\/$/, '') || 'renovate',
    image: cfg.image,
    depName: cfg.depName ?? cfg.image,
    buildArg: cfg.buildArg,
    buildArgs: getArg('build-args', { multi: true }),
    tagSuffix: getArg('tag-suffix') || undefined,
    ignoredVersions: cfg.ignoredVersions ?? [],
    dryRun,
    lastOnly: getInput('last-only') == 'true',
    buildOnly: getInput('build-only') == 'true',
    majorMinor: getArg('major-minor') !== 'false',
    prune: getArg('prune') === 'true',
    versioning: cfg.versioning ?? getDefaultVersioning(cfg.datasource),
  };

  if (dryRun) {
    log('GitHub Actions branch detected - Force building latest, no push');
    config.lastOnly = true;
  }
  log('config:', JSON.stringify(config));

  const token = getArg('token');

  if (token) {
    addHostRule({ matchHost: 'github.com', token });
  }

  await init();

  await generateImages(config);

  log.info(chalk.blue('Processing done:', config.image));
}
