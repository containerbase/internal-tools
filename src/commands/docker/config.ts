import { init } from '../../utils/docker/buildx';

export async function run(): Promise<void> {
  await init();
}
