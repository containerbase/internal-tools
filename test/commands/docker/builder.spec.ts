/* eslint-disable @typescript-eslint/no-var-requires */
import * as _core from '@actions/core';
import * as _datasources from 'renovate/dist/datasource';
import { run } from '../../../src/commands/docker/builder';
import * as _utils from '../../../src/util';
import * as _docker from '../../../src/utils/docker';
import { getName, mocked } from '../../utils';

jest.mock('renovate/dist/datasource');
jest.mock('../../../src/util');
jest.mock('../../../src/utils/docker');
jest.mock('../../../src/utils/docker/buildx', () => ({
  init: () => Promise.resolve(),
}));
jest.mock('../../../src/utils/datasource');

const core = mocked(_core);
const utils = mocked(_utils);
const docker = mocked(_docker);
const datasources = mocked(_datasources);
const version = '1.22.4';

describe(getName(__filename), () => {
  let args: Record<string, string | string[]> = {};
  beforeEach(() => {
    jest.resetAllMocks();
    core.getInput.mockReturnValueOnce('builder.json');
    core.getInput.mockReturnValueOnce('yarn');
    utils.readJson.mockResolvedValueOnce(require('./__fixtures__/yarn.json'));
    args = {};
    utils.getArg.mockImplementation((n, o) => args[n] ?? (o?.multi ? [] : ''));
  });

  it('works yarn', async () => {
    utils.readFile.mockResolvedValue(
      `# renovate: datasource=npm depName=yarn versioning=npm\nARG YARN_VERSION=${version}\n`
    );
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version }, { version: '2.0.0-rc.24' }],
    });

    await run();

    expect(docker.build.mock.calls).toMatchSnapshot('build');
    expect(docker.publish.mock.calls).toMatchSnapshot('publish');
  });

  it('works pnpm', async () => {
    utils.readJson.mockReset();
    utils.readJson.mockResolvedValueOnce(require('./__fixtures__/pnpm.json'));
    args = {
      ...args,
      'image-prefix': 'ghcr.io/renovatebot/',
      'build-args': ['IMAGE=slim'],
      'tag-suffix': 'slim',
    };
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version: '4.0.0-rc.24' }, { version: '5.0.0' }],
    });

    await run();

    expect(docker.build.mock.calls).toMatchSnapshot('build');
    expect(docker.publish.mock.calls).toMatchSnapshot('publish');
  });

  it('works gradle', async () => {
    utils.readJson.mockReset();
    utils.readJson.mockResolvedValueOnce(require('./__fixtures__/gradle.json'));
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [
        { version: '0.7' },
        { version: '2.3' },
        { version: '3.0' },
        { version: '3.5.4' },
        { version: '3.5.5' },
        { version: '4.5' },
        { version: '6.0' },
      ],
    });

    await run();

    expect(docker.build.mock.calls).toHaveLength(3);
    expect(docker.build.mock.calls.map(([args]) => args.tag)).toEqual([
      '3.5.5',
      '4.5',
      '6.0',
    ]);
    expect(docker.build.mock.calls).toMatchSnapshot('build');
    expect(docker.publish.mock.calls).toMatchSnapshot('publish');
  });

  it('works dummy', async () => {
    jest.resetAllMocks();
    utils.readJson.mockResolvedValueOnce(require('./__fixtures__/dummy.json'));

    await run();

    expect(docker.build.mock.calls).toMatchSnapshot('build');
    expect(docker.publish.mock.calls).toMatchSnapshot('publish');
  });

  it('works ubuntu', async () => {
    jest.resetAllMocks();
    utils.readJson.mockResolvedValueOnce(require('./__fixtures__/ubuntu.json'));

    await run();

    expect(docker.build.mock.calls).toMatchSnapshot('build');
    expect(docker.publish.mock.calls).toMatchSnapshot('publish');
  });

  it('last-only dummy', async () => {
    utils.readJson.mockReset();
    utils.readJson.mockResolvedValueOnce(require('./__fixtures__/dummy.json'));
    core.getInput.mockReset();
    core.getInput.mockReturnValueOnce('builder.json');
    core.getInput.mockReturnValueOnce('true');

    args = {
      ...args,
      'build-args': ['IMAGE=slim'],
      'tag-suffix': 'slim',
      'major-minor': 'false',
      prune: 'true',
    };

    await run();

    expect(docker.build.mock.calls).toMatchSnapshot('build');
    expect(docker.publish.mock.calls).toMatchSnapshot('publish');
  });

  it('build-only yarn', async () => {
    core.getInput.mockReturnValueOnce('true');
    core.getInput.mockReturnValueOnce('true');
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version }, { version: '2.0.0-rc.24' }],
    });

    await run();

    expect(docker.build.mock.calls).toMatchSnapshot('build');
    expect(docker.publish.mock.calls).toMatchSnapshot('publish');
  });

  it('updates image yarn (dry-run)', async () => {
    utils.isDryRun.mockReturnValueOnce(true);
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version: '1.5.0' }, { version }],
    });

    await run();

    expect(docker.build.mock.calls).toMatchSnapshot('build');
    expect(docker.publish.mock.calls).toMatchSnapshot('publish');
  });

  it('no releases', async () => {
    datasources.getPkgReleases.mockResolvedValueOnce(null);
    await run();

    expect(docker.build).not.toHaveBeenCalled();
    expect(core.setFailed).toHaveBeenCalledWith('No versions found.');
  });

  it('empty releases', async () => {
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [],
    });

    await run();

    expect(docker.build).not.toHaveBeenCalled();
    expect(core.setFailed).toHaveBeenCalledWith('No versions found.');
  });

  it('unstable releases', async () => {
    utils.readJson.mockReset();
    utils.readJson.mockResolvedValueOnce({
      ...require('./__fixtures__/pnpm.json'),
      ignoredVersions: ['3.5.4'],
    });
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version: '3.5.4' }],
    });

    await run();
    expect(docker.build.mock.calls).toEqual([]);
  });

  it('catch error', async () => {
    expect.assertions(1);
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version }],
    });

    docker.build.mockRejectedValueOnce(new Error('failure'));
    try {
      await run();
    } catch (e) {
      expect((e as Error).message).toEqual('failed');
    }
  });

  it('throws missing-image', async () => {
    expect.assertions(1);
    jest.resetAllMocks();
    utils.readJson.mockResolvedValueOnce({});
    core.getInput.mockReturnValueOnce('');
    core.getInput.mockImplementationOnce(() => {
      throw new Error('missing-image');
    });
    try {
      await run();
    } catch (e) {
      expect((e as Error).message).toEqual('missing-image');
    }
  });
  it('throws missing-config', async () => {
    expect.assertions(1);
    jest.resetAllMocks();
    utils.readJson.mockResolvedValueOnce(undefined);

    try {
      await run();
    } catch (e) {
      expect((e as Error).message).toEqual('missing-config');
    }
  });

  it('multiplatform build-only', async () => {
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version }, { version: '2.0.0-rc.24' }],
    });

    args = {
      ...args,
      platforms: ['linux/amd64', 'linux/arm64'],
    };

    await run();

    expect(docker.build.mock.calls).toMatchSnapshot('build');
    expect(docker.publish.mock.calls).toMatchSnapshot('publish');
  });
});
