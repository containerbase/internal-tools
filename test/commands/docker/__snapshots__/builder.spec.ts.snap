// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build-only yarn: build 1`] = `
Array [
  Array [
    Object {
      "buildArgs": Array [
        "YARN_VERSION=1.22.4",
      ],
      "cache": "docker-build-cache",
      "cacheTags": Array [
        "latest",
        "1",
        "1.22",
      ],
      "dryRun": undefined,
      "image": "yarn",
      "imagePrefix": "renovate",
      "platforms": Array [],
      "tag": "1.22.4",
    },
  ],
]
`;

exports[`build-only yarn: publish 1`] = `Array []`;

exports[`last-only dummy: build 1`] = `
Array [
  Array [
    Object {
      "buildArgs": Array [
        "IMAGE=slim",
        "DUMMY_VERSION=12",
      ],
      "cache": undefined,
      "cacheTags": Array [
        "slim",
      ],
      "dryRun": undefined,
      "image": "dummy",
      "imagePrefix": "renovate",
      "platforms": Array [],
      "tag": "12-slim",
    },
  ],
]
`;

exports[`last-only dummy: publish 1`] = `
Array [
  Array [
    Object {
      "dryRun": undefined,
      "image": "dummy",
      "imagePrefix": "renovate",
      "tag": "12-slim",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "dummy",
      "imagePrefix": "renovate",
      "tag": "slim",
    },
  ],
]
`;

exports[`multiplatform build-only: build 1`] = `
Array [
  Array [
    Object {
      "buildArgs": Array [
        "YARN_VERSION=1.22.4",
      ],
      "cache": "docker-build-cache",
      "cacheTags": Array [
        "latest",
        "1",
        "1.22",
      ],
      "dryRun": undefined,
      "image": "yarn",
      "imagePrefix": "renovate",
      "platforms": Array [
        "linux/amd64",
        "linux/arm64",
      ],
      "tag": "1.22.4",
    },
  ],
]
`;

exports[`multiplatform build-only: publish 1`] = `Array []`;

exports[`updates image yarn (dry-run): build 1`] = `
Array [
  Array [
    Object {
      "buildArgs": Array [
        "YARN_VERSION=1.22.4",
      ],
      "cache": "docker-build-cache",
      "cacheTags": Array [
        "latest",
        "1",
        "1.22",
      ],
      "dryRun": true,
      "image": "yarn",
      "imagePrefix": "renovate",
      "platforms": Array [],
      "tag": "1.22.4",
    },
  ],
]
`;

exports[`updates image yarn (dry-run): publish 1`] = `
Array [
  Array [
    Object {
      "dryRun": true,
      "image": "yarn",
      "imagePrefix": "renovate",
      "tag": "1.22.4",
    },
  ],
  Array [
    Object {
      "dryRun": true,
      "image": "yarn",
      "imagePrefix": "renovate",
      "tag": "1",
    },
  ],
  Array [
    Object {
      "dryRun": true,
      "image": "yarn",
      "imagePrefix": "renovate",
      "tag": "1.22",
    },
  ],
  Array [
    Object {
      "dryRun": true,
      "image": "yarn",
      "imagePrefix": "renovate",
      "tag": "latest",
    },
  ],
]
`;

exports[`works dummy: build 1`] = `
Array [
  Array [
    Object {
      "buildArgs": Array [
        "DUMMY_VERSION=10",
      ],
      "cache": undefined,
      "cacheTags": Array [
        "latest",
      ],
      "dryRun": undefined,
      "image": "dummy",
      "imagePrefix": "renovate",
      "platforms": undefined,
      "tag": "10",
    },
  ],
  Array [
    Object {
      "buildArgs": Array [
        "DUMMY_VERSION=13",
      ],
      "cache": undefined,
      "cacheTags": Array [
        "latest",
      ],
      "dryRun": undefined,
      "image": "dummy",
      "imagePrefix": "renovate",
      "platforms": undefined,
      "tag": "13",
    },
  ],
]
`;

exports[`works dummy: publish 1`] = `
Array [
  Array [
    Object {
      "dryRun": undefined,
      "image": "dummy",
      "imagePrefix": "renovate",
      "tag": "10",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "dummy",
      "imagePrefix": "renovate",
      "tag": "13",
    },
  ],
]
`;

exports[`works dummyx: build 1`] = `
Array [
  Array [
    Object {
      "buildArgs": Array [
        "DUMMY_VERSION=10",
      ],
      "cache": undefined,
      "cacheTags": Array [
        "latest",
      ],
      "dryRun": undefined,
      "image": "dummy",
      "imagePrefix": "renovate",
      "platforms": undefined,
      "tag": "10",
    },
  ],
  Array [
    Object {
      "buildArgs": Array [
        "DUMMY_VERSION=13",
      ],
      "cache": undefined,
      "cacheTags": Array [
        "latest",
      ],
      "dryRun": undefined,
      "image": "dummy",
      "imagePrefix": "renovate",
      "platforms": undefined,
      "tag": "13",
    },
  ],
]
`;

exports[`works dummyx: publish 1`] = `
Array [
  Array [
    Object {
      "dryRun": undefined,
      "image": "dummy",
      "imagePrefix": "renovate",
      "tag": "10",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "dummy",
      "imagePrefix": "renovate",
      "tag": "13",
    },
  ],
]
`;

exports[`works gradle: build 1`] = `
Array [
  Array [
    Object {
      "buildArgs": Array [
        "GRADLE_VERSION=3.5.5",
      ],
      "cache": undefined,
      "cacheTags": Array [
        "latest",
        "3",
        "3.5",
      ],
      "dryRun": undefined,
      "image": "gradle",
      "imagePrefix": "renovate",
      "platforms": Array [],
      "tag": "3.5.5",
    },
  ],
  Array [
    Object {
      "buildArgs": Array [
        "GRADLE_VERSION=4.5",
      ],
      "cache": undefined,
      "cacheTags": Array [
        "latest",
        "4",
      ],
      "dryRun": undefined,
      "image": "gradle",
      "imagePrefix": "renovate",
      "platforms": Array [],
      "tag": "4.5",
    },
  ],
  Array [
    Object {
      "buildArgs": Array [
        "GRADLE_VERSION=6.0",
      ],
      "cache": undefined,
      "cacheTags": Array [
        "latest",
        "6",
      ],
      "dryRun": undefined,
      "image": "gradle",
      "imagePrefix": "renovate",
      "platforms": Array [],
      "tag": "6.0",
    },
  ],
]
`;

exports[`works gradle: publish 1`] = `
Array [
  Array [
    Object {
      "dryRun": undefined,
      "image": "gradle",
      "imagePrefix": "renovate",
      "tag": "3.5.5",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "gradle",
      "imagePrefix": "renovate",
      "tag": "3",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "gradle",
      "imagePrefix": "renovate",
      "tag": "3.5",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "gradle",
      "imagePrefix": "renovate",
      "tag": "4.5",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "gradle",
      "imagePrefix": "renovate",
      "tag": "4",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "gradle",
      "imagePrefix": "renovate",
      "tag": "6.0",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "gradle",
      "imagePrefix": "renovate",
      "tag": "6",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "gradle",
      "imagePrefix": "renovate",
      "tag": "latest",
    },
  ],
]
`;

exports[`works pnpm: build 1`] = `
Array [
  Array [
    Object {
      "buildArgs": Array [
        "IMAGE=slim",
        "PNPM_VERSION=5.0.0",
      ],
      "cache": undefined,
      "cacheTags": Array [
        "slim",
        "5-slim",
        "5.0-slim",
      ],
      "dryRun": undefined,
      "image": "pnpm",
      "imagePrefix": "ghcr.io/renovatebot",
      "platforms": Array [],
      "tag": "5.0.0-slim",
    },
  ],
]
`;

exports[`works pnpm: publish 1`] = `
Array [
  Array [
    Object {
      "dryRun": undefined,
      "image": "pnpm",
      "imagePrefix": "ghcr.io/renovatebot",
      "tag": "5.0.0-slim",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "pnpm",
      "imagePrefix": "ghcr.io/renovatebot",
      "tag": "5-slim",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "pnpm",
      "imagePrefix": "ghcr.io/renovatebot",
      "tag": "5.0-slim",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "pnpm",
      "imagePrefix": "ghcr.io/renovatebot",
      "tag": "slim",
    },
  ],
]
`;

exports[`works ubuntu: build 1`] = `
Array [
  Array [
    Object {
      "buildArgs": Array [
        "UBUNTU_VERSION=bionic",
      ],
      "cache": undefined,
      "cacheTags": Array [
        "latest",
        "18",
        "18.4",
      ],
      "dryRun": undefined,
      "image": "ubuntu",
      "imagePrefix": "renovate",
      "platforms": undefined,
      "tag": "bionic",
    },
  ],
  Array [
    Object {
      "buildArgs": Array [
        "UBUNTU_VERSION=focal",
      ],
      "cache": undefined,
      "cacheTags": Array [
        "latest",
        "20",
        "20.4",
      ],
      "dryRun": undefined,
      "image": "ubuntu",
      "imagePrefix": "renovate",
      "platforms": undefined,
      "tag": "focal",
    },
  ],
]
`;

exports[`works ubuntu: publish 1`] = `
Array [
  Array [
    Object {
      "dryRun": undefined,
      "image": "ubuntu",
      "imagePrefix": "renovate",
      "tag": "bionic",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "ubuntu",
      "imagePrefix": "renovate",
      "tag": "18",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "ubuntu",
      "imagePrefix": "renovate",
      "tag": "18.4",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "ubuntu",
      "imagePrefix": "renovate",
      "tag": "focal",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "ubuntu",
      "imagePrefix": "renovate",
      "tag": "20",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "ubuntu",
      "imagePrefix": "renovate",
      "tag": "20.4",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "ubuntu",
      "imagePrefix": "renovate",
      "tag": "latest",
    },
  ],
]
`;

exports[`works yarn: build 1`] = `
Array [
  Array [
    Object {
      "buildArgs": Array [
        "YARN_VERSION=1.22.4",
      ],
      "cache": "docker-build-cache",
      "cacheTags": Array [
        "latest",
        "1",
        "1.22",
      ],
      "dryRun": undefined,
      "image": "yarn",
      "imagePrefix": "renovate",
      "platforms": Array [],
      "tag": "1.22.4",
    },
  ],
]
`;

exports[`works yarn: publish 1`] = `
Array [
  Array [
    Object {
      "dryRun": undefined,
      "image": "yarn",
      "imagePrefix": "renovate",
      "tag": "1.22.4",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "yarn",
      "imagePrefix": "renovate",
      "tag": "1",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "yarn",
      "imagePrefix": "renovate",
      "tag": "1.22",
    },
  ],
  Array [
    Object {
      "dryRun": undefined,
      "image": "yarn",
      "imagePrefix": "renovate",
      "tag": "latest",
    },
  ],
]
`;
